src/Symfony/Component/Cache/PhpMarshaller/PhpMarshaller.php

```diff
@@ -0,0 +1,158 @@
+            }
+        }
+        $properties = array();
+        foreach ($values as $i => $vars) {
+            foreach ($vars as $class => $values) {
+                foreach ($values as $name => $v) {
+                    $properties[$class][$name][$i] = $v;
+                }
+            }
+        }
+        if (!$classes) {
+            return $value[0];
+        }
+
+        return new Configurator(new Registry($classes), $properties, $value[0], $wakeups);
+    }
+
+    public static function optimize(string $exportedValue)
+    {
+        return preg_replace(sprintf("{%s::__set_state\(array\(\s++'0' => (\d+),\s++\)\)}", preg_quote(Reference::class)), Registry::class.'::$objects[$1]', $exportedValue);
+    }
+
+    private static function doMarshall(array &$array, \SplObjectStorage $objectsPool): int
+    {
+        $objectsCount = 0;
+
+        foreach ($array as &$value) {
+            if (\is_array($value) && $value) {
+                $objectsCount += self::doMarshall($value, $objectsPool);
+            }
+            if (!\is_object($value)) {
+                continue;
+            }
+            if (isset($objectsPool[$value])) {
+                $value = new Reference($objectsPool[$value][0]);
+                continue;
+            }
+            $class = \get_class($value);
+            $properties = array();
+            $sleep = null;
+            $arrayValue = array();
+
+            if ('ArrayIterator' === $class || 'ArrayObject' === $class || 'RecursiveArrayIterator' === $class) {
+                $properties = array(
+                    $value->getArrayCopy(),
+                    $value->getFlags(),
+                    'ArrayObject' === $class ? $value->getIteratorClass() : 'ArrayIterator',
+                );
+                if (array(array(), 0, 'ArrayIterator') === $properties) {
+                    $properties = array();
+                } else {
+                    if ('ArrayIterator' === $properties[2]) {
+                        unset($properties[2]);
+                        if (0 === $properties[1]) {
+                            unset($properties[1]);
+                        }
+                    }
+                    $properties = array($class => array("\0" => $properties));
+                }
+            } elseif ($value instanceof \SplObjectStorage) {
+                foreach (clone $value as $v) {
+                    $properties[] = $v;
+                    $properties[] = $value[$v];
+                }
+                $properties = array('SplObjectStorage' => array("\0" => $properties));
+                $arrayValue = (array) $value;
+            } elseif ($value instanceof \Serializable) {
+                ++$objectsCount;
+                $objectsPool[$value] = array(\count($objectsPool), serialize($value), array(), false);
+                continue;
+            } else {
+                $arrayValue = (array) $value;
+            }
+
+            if (\method_exists($class, '__sleep')) {
```

Comment:
do __sleep and __wakeup methods should be called for classes that implements __set_state?

Reply:
__set_state() is not called anymore so yes, this is required. It's not called because it's not part of what serialize() does, and this code tries to match it.

Reply:
great. so my objects, that expects call to __set_state on restore from opcache is now unusable? It should be noted in docs. Cause __sleep and __wakeup is part of Serialze contract, but __set_state is part of var_export contract.

Reply:
The fact that we use `var_export()` is an internal detail to me: the only thing that PSR-6 mentions is `serialize()`. I wouldn't expect ppl to expect `__set_state()` to be called (it's never been the case), so not sure about the docs, as this might add more confusion actually I fear...

Reply:
I have custom opcache implementation that has serializer as simple as `var_export' (PhpFiles was not be able to opcache objects). It was achieved by implementing __set_state method for objects that going to be cached. How should I migrate to PhpFiles? As I can't override serializer, the only option left is to check every class against non-trivial logic in __set_state and move it to __wakeup method.

Reply:
I'd say this hints a too tight coupling between the things you cache and the storage itself I suppose.
My advice would be do just drop that part...